### PR TITLE
argmax for half datatype fix

### PR DESCRIPTION
**Pull request by pbelevich**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#28787 argmax for half datatype fix**

Differential Revision: [D18194420](https://our.internmc.facebook.com/intern/diff/D18194420)

Fixing #28623

### CODE REVIEW

File: test/test_torch.py [hunk 1]
```diff
@@ -8272,6 +8272,8 @@ def test_dim_reduction(self, device):
                  torch.int64,
                  torch.int32,
                  torch.int16]
+        if device.startswith('cuda'):  # 'cpu' and 'xla' do not support half
```
Comment (SsnL) [inline test/test_torch.py]: `device.type == 'cuda'` is the more idiomatic way to test this

Comment (pbelevich) [reply test/test_torch.py]: but device is a string here

File: test/test_torch.py [hunk 2]
```diff
@@ -8272,6 +8272,8 @@ def test_dim_reduction(self, device):
                  torch.int64,
                  torch.int32,
                  torch.int16]
+        if torch.device(device).type == 'cuda':  # 'cpu' and 'xla' do not support half
```
Comment (ngimel) [inline test/test_torch.py]: `self.device_type == 'cuda':`

**Review by ngimel — APPROVED** [review]
subject to minor change in test_torch

File: test/test_torch.py (hunk 2 above)
Comment (pbelevich) [reply test/test_torch.py]: thanks!

File: aten/src/ATen/native/cuda/ReduceOpsKernel.cu [hunk 3]
```diff
@@ -169,21 +169,37 @@ void min_values_kernel_cuda(TensorIterator& iter) {
...
+void argmin_kernel_cuda_impl(TensorIterator& iter) {
+  gpu_reduce_kernel<scalar_t, int64_t>(
+    iter,
+    ArgMinOps<acc_t>{},
+    thrust::pair<acc_t, int64_t>(at::numeric_limits<acc_t>::upper_bound(), 0));
+};
+
 void argmax_kernel_cuda(TensorIterator& iter) {
+  if (iter.dtype(1) == kHalf) {
+    argmax_kernel_cuda_impl<at::Half, float>(iter);
```
Comment (gchanan) [inline aten/src/ATen/native/cuda/ReduceOpsKernel.cu]: worth a comment to describe why are you doing it this way, since max is easily computable for half.

**Review by gchanan — APPROVED** [review]

File: aten/src/ATen/native/cuda/ReduceOpsKernel.cu (hunk 3 above)
Comment (pbelevich) [reply aten/src/ATen/native/cuda/ReduceOpsKernel.cu]: done

**Pull request closed without merging**